static/_posts/2024-04-01-vgaterm-intro-001.md

```diff
@@ -0,0 +1,171 @@
+<span style="color: {{ site.caption_color }};">Vgaterm Timing circuit top</span> | <span style="color: {{ site.caption_color }};">Vgaterm Timing cicuit bottom</span>
+
+VGA timing is controlled by the "timing board" circuit. This establishes the `H_SYNC` and `V_SYNC` signals that VGA requires, the `START` signal which alerts the CPU to begin  emitting pixels to the FIFO, and the `visible` flag, which is used to compute `CLK_VIS` to read pixel bytes out of the FIFO.
+
+### Pixels Out
+Every frame (60 times a second) the esp32c3 microcontroller needs to output 256,000 bytes (640x400 pixels, 1 byte per pixel) to hardware so they may be displayed. We decided that this is the most important software role as too much jitter in timing will cause artifacts and wobbling in the piture.
+
+To cross the clock domains, from the CPU SPI to VGA pixel clock, we used a FIFO memory chip. This lets us write bytes from the CPU while VGA hardware reads bytes exactly when it needs them.
+
+{% include figure.html image="/assets/images/fifo-chip-highlight.jpg" width=450 position="right" caption="IDT 7205 FIFO chip" %}
+
+This worked well, but there are a few constraints we had to consider when designing a circuit around the FIFO:
+* The FIFO read and write setup and hold times had to be obeyed by `CLK_VIS` and the SPI clock signals respectively
+* The SPI only supported clock speeds in power-of-two divisions of 80 MHz (80, 40, 20, 10, etc)
+* If we emitted pixels too quickly, and VGA hardware could not read from the FIFO fast enough, we would blow the queue and drop pixels if the CPU didn't pause writing periodically.
+* If we emitted pixels too slowly, and VGA hardware consumed pixels faster than was being written, then we'd have to start loading pixels before VGA started consuming them.
+* The size of the FIFO effects how much wiggle room we get in the above two cases, especially as the FIFO is too expensive to have enough room for the whole frame.
+* VGA hardware does not read from the FIFO at the pixel clock rate across the entire duration of the frame. Since the visible portion of the frame is only about 73% of the total time spent per frame (because of the blanking area), the average read rate per frame is actually less than the pixel clock rate
+
+The details in how we eneded up on our specific design will be elaberated on in another post, but suffice to say we had to coordinate design details across software, hardware, specific part choices, and part pricing.
+
+### Displaying Pixels
+To display pixels, the `CLK_VIS` signal generated by the VGA timing hardware reads bytes from the FIFO and into a [Programmable Logic Device (PLD)](https://en.wikipedia.org/wiki/Programmable_logic_device) which maps the 8 bits into 9 bits, 3 bits for each of the color channels - red, green, and blue. These 9 bits then get sent to the DAC, described in the following section.
```

Comment:
```suggestion
To display pixels, the `CLK_VIS` signal generated by the VGA timing hardware triggers reading bytes from the FIFO and into a [Programmable Logic Device (PLD)](https://en.wikipedia.org/wiki/Programmable_logic_device) which maps the 8 bits into 3 bits for each of the color channels - red, green, and blue. These 9 bits then get sent to the DAC, described in the following section.
```

two things:

1. it still feels really weird to me to describe clock signals as having agency; they're the clock on the wall that we all use to coordinate, not the thing that makes us do stuff. Also, did we literally end up wire the clock signal into /R on the FIFO? if so, that feels more like "clever optimization"/chance than agency, still
2. using a whole chip to map "8 bits into 9 bits" is a bit confusing as to why; saying "into 3 bits of each color" -> "oh that happens to be 9 bits" avoids the hiccup in the sentence